### ISSUE TITLE
data race in request and response

### ISSUE THREAD

**Issue by tracyshen** [valyala/fasthttp#576]
**env:**
centos 7.5, golang 1.11.5

**description:**
1. In my project, fasthttp is used as the crawler's httpclient. I found that in the case of high concurrency, the requested url does not match the response.

2. In most of the code, I use `AcquireRequest()` `defer ReleaseRequest()`, `AcquireResponse()` `defer ReleaseResponse()`; A small part of the code is requested using client.GetTimeout().

3. build with -race, I found some data race as below:
![race01](https://user-images.githubusercontent.com/6285163/58085442-52ea5580-7bef-11e9-9876-ee8cd8874bf9.png)
![race02](https://user-images.githubusercontent.com/6285163/58085446-554caf80-7bef-11e9-9cc2-3632936089b0.png)

**Comment by dgrr** [comment]
The race traceback does not make sense to me. Which is the last commit of the fasthttp you are using?

**Comment by erikdubbelboer** [comment]
Are you 100% sure you aren't using the same `Request` multiple times? It's hard to follow the code without knowing which commit you are using but from what I can see in the top trace we have one `Request` being used in code that is running both `doNonNilReqResp` and `clientDoDeadline` which could never be running at the same time unless you reuse a `Request` for 2 calls.

**Comment by tracyshen** [comment]
> The race traceback does not make sense to me. Which is the last commit of the fasthttp you are using?

I tried commit 46469b, same problem

**Comment by tracyshen** [comment]
> Are you 100% sure you aren't using the same `Request` multiple times? It's hard to follow the code without knowing which commit you are using but from what I can see in the top trace we have one `Request` being used in code that is running both `doNonNilReqResp` and `clientDoDeadline` which could never be running at the same time unless you reuse a `Request` for 2 calls.

I'm sure not using the same `Request` at the same same time........  my code as below....

```golang
httpRequest := fasthttp.AcquireRequest()
defer fasthttp.ReleaseRequest(httpRequest)

httpResponse := fasthttp.AcquireResponse()
defer fasthttp.ReleaseResponse(httpResponse)

for {
    if httpRequest.ConnectionClose() {
        httpClient := newFastHTTPClient()
        err = httpClient.DoTimeout(httpRequest, httpResponse, 20*time.Second)
    } else {
        err = crawler.httpClient.DoTimeout(httpRequest, httpResponse, 20*time.Second)
    }

    if err == fasthttp.ErrConnectionClosed && !httpRequest.ConnectionClose() {
        crawler.connectionClose = true
        httpRequest.SetConnectionClose()
        continue
    }

    // ...some code
}
```

**Comment by dgrr** [comment]
Use `request.Reset()` and `response.Reset()` on the top of the for:

**Comment by erikdubbelboer** [comment]
I think I found the issue. I'm trying to fix it.

**Comment by erikdubbelboer** [comment]
I just pushed a fix. @tracyshen can you see if you still have this issue in the latest master version?

**Comment by scarbo87** [comment]
@erikdubbelboer I'm have the problem now.I tried commit 1ee3b845, same problem:

```
WARNING: DATA RACE
Write at 0x00c000c6e0c0 by goroutine 116:
  runtime.slicecopy()
      /usr/local/Cellar/go/1.12.1/libexec/src/runtime/slice.go:197 +0x0
  ..vendor/github.com/valyala/fasthttp.(*RequestHeader).parseHeaders()
      ..vendor/github.com/valyala/fasthttp/header.go:1901 +0xce0
  ..vendor/github.com/valyala/fasthttp.(*RequestHeader).peek()
      ..vendor/github.com/valyala/fasthttp/header.go:1964 +0x7b3
  ..vendor/github.com/valyala/fasthttp.(*Request).MayContinue()
      ..vendor/github.com/valyala/fasthttp/http.go:960 +0x77
  ..vendor/github.com/valyala/fasthttp.(*Request).readLimitBody()
      ..vendor/github.com/valyala/fasthttp/http.go:939 +0x92
  ..vendor/github.com/valyala/fasthttp.(*Server).serveConn()
      ..vendor/github.com/valyala/fasthttp/server.go:1890 +0x1c9b
  ..vendor/github.com/valyala/fasthttp.(*Server).serveConn-fm()
      ..vendor/github.com/valyala/fasthttp/server.go:1801 +0x55
  ..vendor/github.com/valyala/fasthttp.(*workerPool).workerFunc()
      ..vendor/github.com/valyala/fasthttp/workerpool.go:212 +0x104
  ..vendor/github.com/valyala/fasthttp.(*workerPool).getCh.func1()
      ..vendor/github.com/valyala/fasthttp/workerpool.go:184 +0x42
```

**Comment by tracyshen** [comment]
> I just pushed a fix. @tracyshen can you see if you still have this issue in the latest master version?

@erikdubbelboer  I tried lastest master version, still have same problem.......

**Comment by xuecai** [comment]
i think request and response is not safe.
DoTimeout have goroutine to send request and timer for deadline .
@tracyshen   in your code, request and response is reuse, so the reac happened

**Comment by erikdubbelboer** [comment]
I haven't found the issue yet. @tracyshen or @scarbo87 would you be willing to send me your source code privately so I can have a look and see if I can find it?

**Comment by tracyshen** [comment]
@erikdubbelboer  Is this email `erik@dubbelboer.com` ?

**Comment by erikdubbelboer** [comment]
@tracyshen yes.

**Comment by erikdubbelboer** [comment]
@tracyshen are you still having this issue? You haven't replied since my last email.
